captum/optim/_core/objectives.py

```diff
@@ -0,0 +1,261 @@
+
+from contextlib import suppress
+from typing import Callable, Iterable, List, Optional
+
+import torch
+import torch.nn as nn
+import torch.optim as optim
+from tqdm.auto import tqdm
+
+from .._param.image.images import ImageParameterization, NaturalImage
+from .._param.image.transform import RandomAffine
+from .._utils.typing import (
+    LossFunction,
+    ModuleOutputMapping,
+    Objective,
+    Parameterized,
+    SingleTargetLossFunction,
+    StopCriteria,
+)
+from .output_hook import AbortForwardException, ModuleOutputsHook
+
+
+class InputOptimization(Objective, Parameterized):
+    model: nn.Module
+    input_param: ImageParameterization
+    input_transformation: nn.Module
+    """
+    Core function that optimizes an input to maximize a target (aka objective).
+    This is similar to gradient-based methods for adversarial examples, such
+    as FGSM. The code for this was based on the implementation by the authors of Lucid.
+    For more details, see the following:
+        https://github.com/tensorflow/lucid
+        https://distill.pub/2017/feature-visualization/
+    """
+
+    def __init__(
+        self,
+        model: nn.Module,
+        input_param: Optional[nn.Module],
+        transform: Optional[nn.Module],
+        targets: Iterable[nn.Module],
+        loss_function: LossFunction,
+    ):
+        r"""
+        Args:
+            model (nn.Module):  The reference to PyTorch model instance.
+            input_param (nn.Module, optional):  A module that generates
+                        an input, consumed by the model.
+            transform (nn.Module, optional):  A module that transforms or
+                        preprocess the input before being passed to the model.
+            targets (iterable of nn.Module):  A list of targets, objectives
+                        that are used to compute the loss function.
+            loss_function (callable): The loss function to minimize during
+                        optimization.
+        """
+        self.model = model
+        self.hooks = ModuleOutputsHook(targets)
+        self.input_param = input_param or NaturalImage((224, 224))
+        self.transform = transform or RandomAffine(scale=True, translate=True)
+        self.loss_function = loss_function
+
+    def loss(self) -> torch.Tensor:
+        r"""Compute loss value for current iteration.
+        Returns:
+            *tensor* representing **loss**:
+            - **loss** (*tensor*):
+                        Size of the tensor corresponds to the targets passed.
+        """
+        image = self.input_param()._t[None, ...]
+
+        if self.transform:
+            image = self.transform(image)
+
+        with suppress(AbortForwardException):
+            _unreachable = self.model(image)  # noqa: F841
+
+        # consume_outputs return the captured values and resets the hook's state
+        module_outputs = self.hooks.consume_outputs()
+        loss_value = self.loss_function(module_outputs)
```

Comment:
Can we pass `targets` to `loss_function` here so that the user doesn't have to pass `targets` to `channel_activation` or `neuron_activation` when initializing it ? Or perhaps we can do that in the constructor ?

Reply:
Yup i agree with this, a little redundant to be passing same targets twice. I'll look into this and figure out a neater implementation in a future PR.